pkg/scheduler/framework/plugins/podtopologyspread/pod_topology_spread.go

```diff
@@ -26,40 +26,128 @@ import (
 
-// Name is the name of the plugin used in the plugin registry and configurations.
-const Name = "PodTopologySpread"
+const (
+	// Name is the name of the plugin used in the plugin registry and configurations.
+	Name = "PodTopologySpread"
+
+	// preFilterStateKey is the key in CycleState to PodTopologySpread pre-computed data.
+	// Using the name of the plugin will likely help us avoid collisions with other plugins.
+	preFilterStateKey = "PreFilter" + Name
+)
 
 // Name returns name of the plugin. It is used in logs, etc.
 func (pl *PodTopologySpread) Name() string {
 	return Name
 }
 
+// preFilterState computed at PreFilter and used at Filter.
+type preFilterState struct {
+	meta *predicates.PodTopologySpreadMetadata
+}
+
+// Clone makes a copy of the given state.
+func (s *preFilterState) Clone() framework.StateData {
+	copy := &preFilterState{
+		meta: s.meta.Clone(),
+	}
+	return copy
+}
+
+// PreFilter invoked at the prefilter extension point.
+func (pl *PodTopologySpread) PreFilter(ctx context.Context, cycleState *framework.CycleState, pod *v1.Pod) *framework.Status {
+	var meta *predicates.PodTopologySpreadMetadata
+	var allNodes []*nodeinfo.NodeInfo
+	var err error
+
+	if allNodes, err = pl.snapshotSharedLister.NodeInfos().List(); err != nil {
+		return framework.NewStatus(framework.Error, fmt.Sprintf("failed to list NodeInfos: %v", err))
+	}
+
+	if meta, err = predicates.GetPodTopologySpreadMetadata(pod, allNodes); err != nil {
+		return framework.NewStatus(framework.Error, fmt.Sprintf("Error calculating podTopologySpreadMetadata: %v", err))
+	}
+
+	s := &preFilterState{
+		meta: meta,
+	}
+	cycleState.Write(preFilterStateKey, s)
+
+	return nil
+}
+
+// PreFilterExtensions returns prefilter extensions, pod add and remove.
+func (pl *PodTopologySpread) PreFilterExtensions() framework.PreFilterExtensions {
+	return pl
+}
+
+// AddPod from pre-computed data in cycleState.
+func (pl *PodTopologySpread) AddPod(ctx context.Context, cycleState *framework.CycleState, podToSchedule *v1.Pod, podToAdd *v1.Pod, nodeInfo *nodeinfo.NodeInfo) *framework.Status {
+	meta, err := getPodTopologySpreadMetadata(cycleState)
+	if err != nil {
+		return framework.NewStatus(framework.Error, err.Error())
+	}
+
+	meta.AddPod(podToAdd, podToSchedule, nodeInfo.Node())
+	return nil
+}
+
+// RemovePod from pre-computed data in cycleState.
+func (pl *PodTopologySpread) RemovePod(ctx context.Context, cycleState *framework.CycleState, podToSchedule *v1.Pod, podToRemove *v1.Pod, nodeInfo *nodeinfo.NodeInfo) *framework.Status {
+	meta, err := getPodTopologySpreadMetadata(cycleState)
+	if err != nil {
+		return framework.NewStatus(framework.Error, err.Error())
+	}
+
+	meta.RemovePod(podToRemove, podToSchedule, nodeInfo.Node())
+	return nil
+}
+
+func getPodTopologySpreadMetadata(cycleState *framework.CycleState) (*predicates.PodTopologySpreadMetadata, error) {
+	c, err := cycleState.Read(preFilterStateKey)
+	if err != nil {
+		return nil, err
```

Comment:
Read returns an error when the key is not found. CA may chose not to run PreFilter at all, and so the entry will not exist, we don't want to error on that.

Reply:
In that case, can they just bypass registering the plugin as a "PreFilter" extension? Then they won't reach here.

Reply:
I see your point now. It explains my concern at https://github.com/kubernetes/kubernetes/pull/86175#discussion_r357010185. Thanks.

Reply:
In this case, I need to throw an error when the meta is nil in Filter(). (Because autoscaler doesn't have logic implemented for the "else" block for even pods spread yet)

Will add the logic.

Reply:
Ah, hold on.

In EvenPodsSpread, it returns nil as a **legit** meta when it's unnecessary to proceed pre-computing, e.g., no spread constraints in the pod, etc. But for sure, the nil meta will still be written to `CycleState`.
So if CA chooses not to run PreFilter, then the state entry (wrapping the meta) will be nil, then we will get NotFound error. If we tolerate the NotFound error, and assign the meta to nil, it will get conflicted with its intentional semantics. (In that case, a nil meta will give green light for each Filter() call).

Reply:
Per offline talk with @ahg-g, raised a new commit to adjust the semantic of `nil` "PodTopologySpreadMetadata". It's intentional to keep it as a separate commit in case of reverting in the future.